extensions/amp-video-docking/1.0/dock.js

```diff
@@ -0,0 +1,488 @@
+
+    const transitionDurationMs = 300;
+
+    return this.placeAt_(
+      entry,
+      x,
+      y,
+      scale,
+      step,
+      transitionDurationMs,
+      relativeX,
+      'absolute'
+    ).then(() => {
+      this.current_ = null;
+      this.setStyles_(null);
+    });
+  }
+
+  /**
+   * Returns the area's target when a video should be docked.
+   * @param {!IntersectionObserverEntry} entry
+   * @return {?DockTargetDef}
+   * @private
+   */
+  getTargetFor_(entry) {
+    if (
+      this.current_
+      // this.isDragging_ ||
+      // !this.isValidSize_(video) ||
+      // this.ignoreBecauseAnotherDocked_(video) ||
+      // this.ignoreDueToNotPlayingManually_(video)
+    ) {
+      return null;
+    }
+
+    const {intersectionRect} = entry;
+    if (!isSizedRect(intersectionRect)) {
+      return null;
+    }
+    if (intersectionRect.top > this.getTopBoundary_()) {
+      return null;
+    }
+    return this.getUsableTarget_(entry);
+  }
+
+  /**
+   * @return {number}
+   * @private
+   */
+  getTopBoundary_() {
+    // TODO: Not supporting slot yet.
+    // const slot = this.getSlot_();
+    // if (slot) {
+    //   // Match slot's top edge to tie transition to element.
+    //   return slot.getPageLayoutBox().top;
+    // }
+    return 0;
+  }
+
+  /**
+   * Returns usable target. If it can dock to a `slot` element that's larger
+   * than zero width/height, it calculates a letterboxed rectangle that matches
+   * the aspect ratio of the `video` but is contained within the slot's box.
+   * Otherwise returns a rectangle calculated relative to viewport and sticking
+   * to the horizontal `directionX`.
+   * @param {!IntersectionObserverEntry} entry
+   * @return {!DockTargetDef}
+   * @private
+   */
+  getUsableTarget_(entry) {
+    // TODO(alanorozco): Select slot.
+    return {
+      type: DockTargetType.CORNER,
+      rect: topCornerRect(
+        entry.boundingClientRect,
+        entry.rootBounds,
+        this.cornerDirectionX_,
+        CORNER_WIDTH_RATIO,
+        CORNER_WIDTH_MIN,
+        CORNER_MARGIN_RATIO,
+        CORNER_MARGIN_MAX
+      ),
+      // Bookkeep horizontal edge for drag-and-snap.
+      directionX: this.cornerDirectionX_,
+    };
+  }
+}
+
+export const DockContext = createContext({});
+
+const ShadowLayer = ({styles}) => {
+  const classes = useStyles();
+  return (
+    <div
+      className={`amp-video-docked-shadow ${classes.shadowLayer}`}
+      style={{
+        ...styles?.dockedStyle,
+        'transition-property': 'transform, opacity',
+      }}
+      hidden={!styles}
+    ></div>
+  );
+};
+
+/**
+ * @param {*} props
+ * @return {!PreactDef.Renderable}
+ */
+export function Dock({children}) {
+  const [handle, setHandle] = useState(null);
+  const [state, setState] = useState(null);
+  const [styles, setStyles] = useState(null);
+  const manager = useMemo(
+    () => new DockManager(document, setState, setStyles),
+    [setStyles]
```

Comment:
Question for my own knowledge: Why is `setStyles` a dependency but `setState` isn't?